src/development/containerisation-standards.rst

```diff
@@ -0,0 +1,630 @@
+* Volumes - persistent data and configuration files
+* Ports
+* Environment variables
+* Permissions
+* Memory
+* CPU
+* Devices
+* OS tuning, and ulimits
+* IPC
+* Signal handling
+* Command and arguments
+* Treatment of StdIn, StdOut, and StdErr
+
+Usage documentation for the image must describe the intended purpose of each of these configurable resources where consumed, how they combine and what the defaults are with default behaviours.
+
+
+Container Resources
+-------------------
+
+Management of container resources is largely dependent on the specific Container Engine in use.  For example, Docker by default runs a container application in it's own namespace as the root user, however this is highly configurable.  The following example shares devices, and user details with the host OS, effectively transparently running the application as the current user of the command line:
+
+.. code:: bash
+
+    cat <<EOF | docker build -t mplayer -
+    FROM ubuntu:18.04
+    ENV DEBIAN_FRONTEND noninteractive
+    RUN \
+        apt update && \
+        apt install mplayer -y && \
+        apt clean && \
+        rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
+
+    ENTRYPOINT ["/usr/bin/mplayer"]
+    CMD ["--help"]
+    EOF
+
+    docker run --rm --name the-morepork-owl \
+      --env HOME=${HOME} \
+      --env DISPLAY=unix$DISPLAY \
+      --volume /etc/passwd:/etc/passwd:ro \
+      --user $(id -u) \
+      --volume ${HOME}:${HOME} \
+      --workdir ${HOME} \
+      --volume /tmp/.X11-unix:/tmp/.X11-unix:ro \
+      --volume /etc/machine-id:/etc/machine-id:ro \
+      --volume /run/user/$(id -u):/run/user/$(id -u):ro \
+      -ti mplayer /usr/bin/mplayer https://www.doc.govt.nz/Documents/conservation/native-animals/birds/bird-song/morepork-song.mp3
+
+
+Storage
+~~~~~~~
+
+As previously stated, all storage shared into a container is achieved through bind mounting.  This is true for both directory mount points and individual files. While it is not mandatory to use the ``VOLUMES`` directive in the image ``Dockerfile``, it is good practice to do this for all directories to be mounted as it provides annotation of the image requirements.
+These volumes and files can be populated with default data, but be aware they are completely masked at runtime when overlayed by a mount.
+
+When adding a volume at runtime, consider whether write access is really required.  As with the example above ``--volume /etc/passwd:/etc/passwd:ro`` ensures that the ``/etc/passwd`` file is read only in the container reducing the security concerns.
+
+
+.. _header-3-network:
+
+Network
+~~~~~~~
+
+containerised applications should avoid using ``--net=host`` (host only) based networking as this will push the container onto the running host network namespace monopolising any ports that it uses.  This means that another instance of this container or any other that uses the same ports cannot run on the same host severely impacting on scheduling and resource utilisation efficiencies.
```

Comment:
Depending on how the code accesses RDMA or any other specialised device, it should be possible to share the device directly into the container using the file handle.  If this is not feasible then the fall back is to use --net=host, but as pointed out, this pushes the container onto the root network namespace meaning that it will monopolise resources.  You can still have multiple containers on the same host using the --net=host, but you now have to track what ports etc. they use to avoid conflict.  You also need to be aware that any container that is launched with --net=host is now privileged too.

What should happen is that infrastructure sorts out the presentation of networks to the host, and then the network device driver integrated with the container engine negotiates access - a bit like macvlan, or how calico enables k8s to drop ip encapsulation by adding the container interface to the host adapter.